src/main/java/com/google/sps/servlets/ProjectServlet.java

```diff
@@ -0,0 +1,266 @@
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.CompositeFilter;
+import com.google.appengine.api.datastore.Query.CompositeFilterOperator;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.users.User;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Handles POST and GET requests for projects,
+ * allows creation and update of projects, and
+ * supports queries for projects based on various parameters.
+ */
+@WebServlet("/projects")
+public class ProjectServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Must be logged in
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/"); // placeholder: should redirect to login
+      return;
+    }
+
+    // Mode is a required parameter
+    boolean isCreateMode = DataUtils.parseMode(request, response);
+
+    String uEmail = userService.getCurrentUser().getEmail();
+    String projId = request.getParameter("proj-id");
+
+    // Will be either a new entity for creation or an existing entity for
+    // updating
+    Entity projEntity = new Entity(DataUtils.PROJECT);
+
+    if (!isCreateMode) {
+      // Must be owner to update
+      projEntity = DataUtils.getProjectEntity(projId, uEmail, false, false);
+
+      // Delete overrides all other updates
+      boolean delete = Boolean.parseBoolean(request.getParameter("delete"));
+      if (delete) {
+        datastore.delete(projEntity.getKey());
```

Comment:
We should also send the user an alert to make sure they actually want to delete the proj, always good to have a safe guard, this would be in the UI though when they click the button, I just wanted to note it and add it to the design doc.